### ISSUE TITLE
I want to partially compare protos

### ISSUE THREAD

**Issue by hickford** [golang/protobuf#1436]
In C++ unit tests with [GoogleTest](https://github.com/google/googletest), I can partially compare protos with `Partially(EqualsProto(...))` ([Google-internal docs](http://go/protomatchers), [open-source fork](https://github.com/google/nucleus/blob/master/nucleus/testing/protocol-buffer-matchers.h)).

> only fields present in the expected protobuf are considered

Is there any similar library for Golang?

**Comment by znkr** [comment]
[protocmp](https://pkg.go.dev/google.golang.org/protobuf/testing/protocmp) is the recommended library to use for proto comparisons. I don't think the same matching functionality exists there though. Maybe create a feature request for that library?

**Comment by hickford** [comment]
>  Maybe create a feature request for that library?

package https://pkg.go.dev/google.golang.org/protobuf/testing/protocmp links to repo https://github.com/protocolbuffers/protobuf-go but that repo has issues disabled

**Comment by znkr** [comment]
Oups. I guess I should have known that :)

**Comment by puellanivis** [comment]
Seems like something easy enough to extend out of [protocmp.IgnoreDefaultScalars](https://pkg.go.dev/google.golang.org/protobuf/testing/protocmp#IgnoreDefaultScalars), literally just a copy-paste and delete maybe three lines.

**Comment by znkr** [comment]
Mhmm, that sounds like it's not too much work indeed. I think the main difference would be that would have to work asymmetrically (at least that's how I understand how partial matching works in C++). I am not sure if that fits well into cmp. An alternative would be to create a filter based on fields set in the message to compare against. Something like `protocmp.IgnoreUnsetFieldsIn(want))`.

**Comment by puellanivis** [comment]
We shouldn’t need to include a pointer to a “unset fields from.” Well, that is unless the `Cmp` is itself intentionally unordered (it does not recommend which is supposed to be the got or expected). In which case, we could also just have `IgnoreFieldsUnsetInLeft()` vs `IgnoreFieldsUnsetInRight()`. This would avoid the need for a duplicated reference to one of the parameters.

**Comment by puellanivis** [comment]
[cmp.FilerPath](https://pkg.go.dev/github.com/google/go-cmp/cmp#FilterPath) also requires it to be symmetric. So, yeah. I think we would have to use a `IgnoreUnsetFieldsFrom(want)` form.

**Comment by znkr** [comment]
Thanks for checking! We're not sure if and when we get to this, but we're definitely happy to review contributions. :)

**Comment by dsnet** [comment]
> [cmp.FilterPath](https://pkg.go.dev/github.com/google/go-cmp/cmp#FilterPath) also requires it to be symmetric. So, yeah. I think we would have to use a IgnoreUnsetFieldsFrom(want) form.

Correct, the `cmp` functionality needs to be symmetric, but it would be permissible to do something like:
```go
cmp.Diff(x, y, protocmp.FilterTemplate(y))
```
where the comparison for `x` and `y` is symmetric, but what parts of the tree that is examine is determined based on the argument `y` passed to a hypothetical `protocmp.FilterTemplate`.

**Comment by torchsec-roni** [comment]
I'd love to have this feature, if it is indeed just a couple lines of diff from [protocmp.IgnoreDefaultScalars](https://pkg.go.dev/google.golang.org/protobuf/testing/protocmp#IgnoreDefaultScalars), I'm happy to put together a PR.

**Comment by torchsec-roni** [comment]
In case this is helpful for anyone, I solved this in the meantime as follows:

```go
// test_utils.go
package test_utils

import (
	"reflect"

	_ "unsafe" // Required for go:linkname

	"github.com/google/go-cmp/cmp"
	"google.golang.org/protobuf/testing/protocmp"
)

var messageReflectType = reflect.TypeOf(protocmp.Message{})

//go:linkname isDefaultScalar google.golang.org/protobuf/testing/protocmp.isDefaultScalar
func isDefaultScalar(m protocmp.Message, k string) bool

func IgnoreDefaultScalarsInLeft() cmp.Option {
	return cmp.FilterPath(func(p cmp.Path) bool {
		// Filter for Message maps.
		mi, ok := p.Index(-1).(cmp.MapIndex)
		if !ok {
			return false
		}
		ps := p.Index(-2)
		if ps.Type() != messageReflectType {
			return false
		}

		// Check whether both fields are default or unpopulated scalars.
		vx, _ := ps.Values()
		mx := vx.Interface().(protocmp.Message)
		k := mi.Key().String()
		return isDefaultScalar(mx, k)
	}, cmp.Ignore())
}
```

Usage in a test:
```go
// mytest.go
func TestSomething(t *testing.T) {
	readUserResp, err := apiServer.ReadUser(ctx, &apiv1.ReadUserRequest{
		Id: "someid",
	})
	assert.NoError(t, err)

	diff := cmp.Diff(&apiv1.User{
		Id: "someid",
		Name: "some name",
	}, readUserResp.User, protocmp.Transform(), test_utils.IgnoreDefaultScalarsInLeft())
	assert.Empty(t, diff, "unexpected user: %s", diff)
}
```
